src/coreclr/vm/comdelegate.cpp

```diff
@@ -1971,11 +1971,12 @@ Stub* COMDelegate::GetInvokeMethodStub(EEImplMethodDesc* pMD)
     {
         // Validate the invoke method, which at the moment just means checking the calling convention
 
-        if (*pMD->GetSig() != (IMAGE_CEE_CS_CALLCONV_HASTHIS | IMAGE_CEE_CS_CALLCONV_DEFAULT))
-            COMPlusThrow(kInvalidProgramException);
-
         MetaSig sig(pMD);
 
+        BYTE callConv = sig.GetCallingConventionInfo();
+        if (callConv != (IMAGE_CEE_CS_CALLCONV_HASTHIS | IMAGE_CEE_CS_CALLCONV_DEFAULT))
+            COMPlusThrow(kInvalidProgramException);
```

Comment:
Just curious, does this use HMF or is it already on the [asm helper](https://github.com/dotnet/runtime/pull/108480#issuecomment-2388670105) plan under the hood?

Reply:
It doesn't use HMF since this must be called when an actual transition is performed - a QCall. The QCall erects a normal transition frame to unmanaged. Note that `SuppressGCTransition` on a QCall makes this call invalid. The `COMPlusThrow` APIs hook into the normal .NET exception mechanism here. I'm not the expert in this domain so if you need more details @jkotas or @janvorli are best to comment on how it aligns with the referenced approach/comment.